lib/World.luau

```diff
@@ -151,36 +125,217 @@ function World:spawnAt(id, ...)
+	if not world:contains(id) then
+		error(ERROR_NO_ENTITY, 2)
 	end
 
 	local components = {}
 	local metatables = {}
+	local entity = world:_getEntity(id)
 
-	for i = 1, select("#", ...) do
-		local newComponent = select(i, ...)
+	for index, component in replaceCommand.components do
+		assertValidComponentInstance(component, index)
 
-		assertValidComponentInstance(newComponent, i)
+		local metatable = getmetatable(component)
+		if components[metatable] then
+			error(("Duplicate component type at index %d"):format(index), 2)
+		end
 
-		local metatable = getmetatable(newComponent)
+		world:_trackChanged(metatable, id, entity[metatable], component)
 
-		if components[metatable] then
-			error(("Duplicate component type at index %d"):format(i), 2)
+		components[metatable] = component
+		table.insert(metatables, metatable)
+	end
+
+	for metatable, component in pairs(entity) do
+		if not components[metatable] then
+			world:_trackChanged(metatable, id, component, nil)
 		end
+	end
+
+	world._entityMetatablesCache[id] = metatables
+	world:_transitionArchetype(id, components)
+end
 
-		self:_trackChanged(metatable, id, nil, newComponent)
+local function executeRemove(world: World, removeCommand: RemoveCommand)
+	local id = removeCommand.entityId
+	if not world:contains(id) then
+		error(ERROR_NO_ENTITY, 2)
+	end
+
+	local entity = world:_getEntity(id)
+
+	local removed = {}
+	for index, metatable in removeCommand.components do
+		assertValidComponent(metatable, index)
+
+		local oldComponent = entity[metatable]
+		removed[index] = oldComponent
+
+		world:_trackChanged(metatable, id, oldComponent, nil)
+		entity[metatable] = nil
+	end
 
-		components[metatable] = newComponent
+	-- Rebuild entity metatable cache
+	local metatables = {}
+	for metatable in pairs(entity) do
 		table.insert(metatables, metatable)
 	end
 
-	self._entityMetatablesCache[id] = metatables
+	world._entityMetatablesCache[id] = metatables
+	world:_transitionArchetype(id, entity)
+end
 
-	self:_transitionArchetype(id, components)
+local function processCommand(world: World, command: Command)
+	if command.type == "spawn" then
+		executeSpawn(world, command)
+	elseif command.type == "insert" then
+		executeInsert(world, command)
+	elseif command.type == "despawn" then
+		executeDespawn(world, command)
+	elseif command.type == "remove" then
+		executeRemove(world, command)
+	elseif command.type == "replace" then
+		executeReplace(world, command)
+	else
+		error(`Unknown command type: {command.type}`)
+	end
+end
+
+local function bufferCommand(world: World, command: Command)
+	if world.deferring then
+		-- Attach current debug context because we don't know when this command will be processed.
+		(command :: any).debugContext = world.debugContext
+		table.insert(world.commands, command)
+	else
+		processCommand(world, command)
+	end
+end
+
+--[=[
+	Starts deferring entity commands.
+
+	If you are using a [`Loop`](/api/Loop), this is done for you.
+]=]
+function World:startDeferring()
```

Comment:
just do self.deferring = true as calling it twice in a row should not be erroneous because the function is idempotent, nor is there any performance penalties. In fact it would be doing less.